packages/@aws-cdk/aws-apprunner/lib/service.ts

```diff
@@ -0,0 +1,722 @@
+import * as ecr from '@aws-cdk/aws-ecr';
+import { DockerImageAsset } from '@aws-cdk/aws-ecr-assets';
+import * as iam from '@aws-cdk/aws-iam';
+import * as cdk from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { CfnService } from './';
+
+/**
+ * The image repository types
+ */
+export enum ImageRepositoryType {
+  /**
+   * Amazon ECR Public
+   */
+  ECR_PUBLIC = 'ECR_PUBLIC',
+
+  /**
+   * Amazon ECR
+   */
+  ECR = 'ECR',
+}
+
+/**
+ * The number of CPU units reserved for each instance of your App Runner service.
+ */
+export class Cpu {
+  /**
+   * 1 vCPU
+   */
+  public static readonly ONE_VCPU = Cpu.of('1 vCPU')
+
+  /**
+   * 2 vCPU
+   */
+  public static readonly TWO_VCPU = Cpu.of('2 vCPU')
+
+  /**
+   * Custom CPU unit
+   *
+   * @param unit custom CPU unit
+   */
+  public static of(unit: string) { return new Cpu(unit); }
+
+  /**
+   *
+   * @param unit The unit of CPU.
+   */
+  private constructor(public readonly unit: string) {}
+}
+
+
+/**
+ * The amount of memory reserved for each instance of your App Runner service.
+ */
+export class Memory {
+  /**
+   * 2 GB
+   */
+  public static readonly TWO_GB = Memory.of('2 GB')
+
+  /**
+   * 3 GB
+   */
+  public static readonly THREE_GB = Memory.of('3 GB')
+
+  /**
+   * 4 GB
+   */
+  public static readonly FOUR_GB = Memory.of('4 GB')
+
+  /**
+   * Custom Memory unit
+   *
+   * @param unit custom Memory unit
+   */
+  public static of(unit: string) { return new Memory(unit); }
+
+  /**
+   *
+   * @param unit The unit of memory.
+   */
+  private constructor(public readonly unit: string) { }
+}
+
+/**
+ * The code runtimes
+ */
+export enum CodeRuntime {
+  /**
+   * NodeJS 12
+   */
+  NODEJS_12 = 'NODEJS_12',
+
+  /**
+   * Python 3
+   */
+  PYTHON_3 = 'PYTHON_3',
+}
+
+/**
+ * The configuration for the Code.
+ */
+export interface CodeConfig {
+  /**
+   * The image repository configuration.
+   *
+   * @default - no image repository.
+   */
+  readonly imageRepository?: ImageRepository;
+
+  /**
+   * The ECR repository.
+   *
+   * @default - no ECR repository.
+   */
+  readonly ecrRepository?: ecr.IRepository;
+
+  /**
+   * The code repository configuration.
+   *
+   * @default - no code repository.
+   */
+  readonly codeRepository?: CodeRepositoryProps;
```

Comment:
Is this a situation where you can use only ONE of those options? Or can them all be used at the same time? If it's a "one in" kind of situation, then these should probably be modelled differently.

Reply:
According to current design, the `source` property from the `Service` construct requires the `CodeConfig` interface(see [here](https://github.com/pahud/aws-cdk/blob/5b9095e203433127ed3f567304e49133dd5ed0cb/packages/%40aws-cdk/aws-apprunner/lib/service.ts#L361)), which we can tell if it's `Code.fromImage()` or `Code.fromGithub()` and continue processing according to the available properties from the `CodeConfig` returned. Not sure if this is a smart design. However, If we model differently then we probably would make  `source: SourceConfig` with `interface CodeConfig extends SourceConfig` and `interface ImageConfig extends SourceConfig` and determine different scenarios between `CodeConfig` and `ImageConfig`.  What do you think?